### ISSUE TITLE
Nested dict interface to choicemaps and traces

### ISSUE THREAD

**Issue by bzinberg** [probcomp/Gen.jl#156]
Quoting https://github.com/probcomp/Gen/issues/106#issuecomment-502407527:
> I think that two separate changes could help with Gen's usability:
> 
> 1. Allow `Base.getindex` on a `Trace` to return a submap, and not just a random choice. (See https://github.com/probcomp/Gen/blob/master/src/gen_fn_interface.jl#L71-L73 and https://github.com/probcomp/Gen/blob/master/src/choice_map.jl#L75.
> 
> 2. Make choice maps iterable, where we iterate over both submaps and choices (i.e. where the iterator is an iterator over pairs of key and value where the value is either a random choice value or a choicemap).
> 
> 
> However, a potential issue with both of these (and the reason why they aren't already present) is that it is technically valid to have random choices that are themselves choice maps (nothing in Gen prevents this). Therefore, there would be a potential ambiguity, where it's not clear whether a value is a random choice or a choice map. It might not matter in the end, but it requires some thought. It also relates to Gen's semantics.

Quoting https://github.com/probcomp/Gen/issues/106#issuecomment-502409233:
> The information about whether a given submap in a choice map is a random choice or a choice map need not be maintained, and this ambiguity need never be resolved.

Quoting https://github.com/probcomp/Gen/issues/106#issuecomment-502410946:
> Yeah, I think that the distinction between the two cases can probably be elided by the choice map interface, which would result in simplified implementations and user interfaces for choice maps. I would want to think through the implications for JIT-generation in the static modeling language (I don't think there would be issues, but not 100% sure). Also a lot of code would be touched (and in general, simplified I think).

Given the above, I think it makes sense to:
1. Add a wrapper like the one implemented [here](https://gist.github.com/bzinberg/89253afc30de37a628b7c9fbdcdeb751/a3b5f11dc8c32729864410be4885a2887664dcac) to Gen, available via a short function name, for example
```julia
c(trace)[key][subkey]  # equivalent to `trace[key => subkey]`
```

2. Keep the heavier-weight engineering work of designing and implementing any (possibly breaking) changes to the choicemap interface, and implementation of the above improvements/simplifications, in the issue backlog for when we have time to tackle that part.

**Comment by bzinberg** [comment]
@marcoct, if the above sounds good to you, I can send a PR for 1 above (the wrapper version).  Wdyt?

**Comment by alex-lew** [comment]
@bzinberg Hm, my gut reaction is that we should avoid introducing single-letter method names; is there a more informative name for the functionality you want? If you really want something short, you can always do `const c = Gen.longerName` in your own project. (In general, if we are planning on editing the choicemap interface to support this without a function `c` and then deprecate this function, I wonder if it makes sense to include this modularly as a separate package / as a file in projects that use it, rather than in Gen itself. Of course, Gen is rapidly changing anyway, so maybe it’s fine.)

**Comment by bzinberg** [comment]
Yeah, I'm happy to call this something like `Gen.nested_dict_view` and anyone who wants to alias it can alias it (cf. `import numpy as np`).

I think this is the kind of thing that would go in "contrib" or "staging" for a production software product.  But as you say, I think it would be fine to add this now and deprecate it later, given the existing loud warnings that Gen is rapidly evolving research software (which we currently under-utilize, IMO :slightly_smiling_face:).  Wdyt?
